File: src/main/java/gr/apt/controller/PersonController.java
package gr.apt.controller;

import gr.apt.exception.LmsException;
import gr.apt.persistence.dto.PersonDto;
import gr.apt.service.PersonService;

import javax.inject.Inject;
import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import java.math.BigInteger;
import java.util.List;

@Path("/person")
@Produces("application/json")
public class PersonController {
    @Inject
    PersonService service;

    @GET
    @Path("/list")
    public Response findAll(@QueryParam("index") Integer index, @QueryParam("size") Integer size) throws LmsException {
        List list = service.findAll(index,size);
        if(list != null){
            return Response.ok(list).build();
        }
        return Response.status(Response.Status.CONFLICT).build();
    }

    @GET
    @Path("/{id}")
    public Response findById(@PathParam("id") BigInteger id) throws LmsException {
        PersonDto dto = service.findById(id);
        if(dto != null){
            return Response.ok(dto).build();
        }
        throw new LmsException("Could not find Leave with id:"+id);
    }

    @POST
    @Consumes("application/json")
    public Response create(PersonDto dto) throws LmsException {
        return service.create(dto) ? Response.ok(dto).build():Response.status(Response.Status.CONFLICT).build();
    }

    @PUT
    @Consumes("application/json")
    public Response update(PersonDto dto) throws LmsException {
        return service.update(dto) ? Response.ok(dto).build():Response.status(Response.Status.CONFLICT).build();
    }

    @DELETE
    @Consumes("application/json")
    public Response delete(PersonDto dto) throws LmsException {
        return service.delete(dto) ? Response.ok().build():Response.status(Response.Status.CONFLICT).build();
    }

    @GET
    @Path("/refresh")
    public Response refresh() throws LmsException {
        return service.refreshDaysOfLeaves() ? Response.ok().build():Response.status(Response.Status.CONFLICT).build();
    }
}

File: src/main/java/gr/apt/lms/controller/PersonController.kt
package gr.apt.lms.controller

import gr.apt.lms.dto.person.PersonDto
import gr.apt.lms.exception.LmsException
import gr.apt.lms.service.PersonService
import java.math.BigInteger
import javax.inject.Inject
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path("/person")
@Produces(MediaType.APPLICATION_JSON)
class PersonController {

    @get:Inject
    lateinit var service: PersonService

    @GET
    @Path("/list")
    @Throws(LmsException::class)
    fun findAll(@QueryParam("index") index: Int?, @QueryParam("size") size: Int?) =
        Response.ok(service.findAll(index, size)).build()


    @GET
    @Path("/{id}")
    @Throws(LmsException::class)
    fun findById(@PathParam("id") id: BigInteger) = Response.ok(service.findById(id)).build()


    @POST
    @Consumes(MediaType.APPLICATION_JSON)
    @Throws(LmsException::class)
    fun create(dto: PersonDto) = Response.ok(service.create(dto)).build()

    @PUT
    @Consumes(MediaType.APPLICATION_JSON)
    @Throws(LmsException::class)
    fun update(dto: PersonDto) = Response.ok(service.update(dto)).build()

    @DELETE
    @Consumes(MediaType.APPLICATION_JSON)
    @Throws(LmsException::class)
    fun delete(dto: PersonDto) = Response.ok(service.delete(dto)).build()

    //TODO: replace this endpoint with a cronjob if possible
    @GET
    @Path("/refresh")
    @Throws(LmsException::class)
    fun refresh() = Response.ok(service.refreshDaysOfLeaves()).build()
}